### PR TITLE
Platform admin billing report

### PR DESCRIPTION
**Pull request by leohemsted**
The platform admin billing report is different to the other two billing reports in one crucial way: It allows you to report on a period of time that doesn't need to be an entire financial year. This means that it needs to be aware of the part of the financial year leading up to the billing period - which it _shouldn't_ include in the results, but it does need to know how many fragments were already spent.

Fortunately we were able to re-use a lot of the existing query code from the organisation usage report. That already has some tests for the complicated subquery magic, so we don't need to re-test them.

However, we ended up having to modify the existing code a bit - i need to have another look now that i've got the existing tests passing to see if there's any sensible refactors/cleanups because I feel like we've got a lot of columns that look very similar but slightly different, so any suggestions that people have i'm keen to hear.

(Also, better ways of testing this than deploying to prod and comparing the sms report results is also welcome)

### CODE REVIEW

**Review by benthorner — COMMENTED** [review]
I think this is almost good to go, although I found it a hard one to review:

- Some of the changes aren't captured in the commit messages.
- The third commit introduces a bug that's fixed by the fourth.

Ideally some of the changes could have gone in smaller commits: the `coalesce`s and the removal of `rate` would have been good candidates and are the reason for me not approving: I want to double check they're both intentional changes.

## Testing

I found the existing tests hard to follow as they operate at multiple levels:

1. One test of the top-level API (good).
2. Tests of the DAO function (ok).
3. Tests of the lower-level queries (err).

In contrast, `fetch_billing_totals_for_year` only has two layers of testing, with the bulk of the tests at the second layer. Higher-level tests provide more coverage and make it easier to check the extent of coverage, so I've made a few comments towards that.

In general though I think we've basically covered all the critical cases:

- Usage in and before the specified period.
- Free allowance before specific period (test has [redundant data](https://github.com/alphagov/notifications-api/blob/43dbc0891fc0db8485bdfc6c7482fdbe85e7bf1d/tests/app/dao/test_fact_billing_dao.py#L696)).
- Free allowance used up during specified period.
- Multiple rates (lower level only).
- Missing: usage after the specified period.
- Missing: no data for a service (no row in results).
- Missing: trial services are excluded (implicit in ☝️).
- Missing: calculation takes rate_multiplier into account.

None of the changes here significantly affect any of the missing cases above, so not a reason to block this PR but would be good to look at if we have time.

## Stretch goal

Currently we have two very similar subqueries to computing billing data, with one pointing at the other for explanation. This is actually quite confusing as they both work differently - the joins in particular - and the variable names in the comment are out-of-date.

I had a quick go to see if it's possible to combine them and I think it is: https://github.com/alphagov/notifications-api/compare/plat-adm-billing-report...ben-test-combine-queries?expand=1

---

Happy to work on some of this together and look through the comments. I realise there are quite a lot of comments. The complexity of this stuff - and all the very similar function names 🤯 - makes it a hard one to do a more informal review straight off.

File: app/dao/fact_billing_dao.py [hunk 1]
```diff
@@ -64,53 +64,52 @@ def fetch_sms_free_allowance_remainder_until_date(end_date):
...
 def fetch_sms_billing_for_all_services(start_date, end_date):
+    # ASSUMPTION: start and end date are in the same financial year
+    financial_year = get_financial_year_for_datetime(get_london_midnight_in_utc(start_date))
 
     # ASSUMPTION: AnnualBilling has been populated for year.
-    allowance_left_at_start_date_query = fetch_sms_free_allowance_remainder_until_date(start_date).subquery()
+    ft_billing_subquery = query_sms_usage_for_year_per_service(financial_year).subquery()
 
-    sms_billable_units = func.sum(FactBilling.billable_units * FactBilling.rate_multiplier)
+    sms_billable_units = func.sum(ft_billing_subquery.c.chargeable_units)
```
[the remark this answers is not in the document]
Comment (leohemsted) [reply app/dao/fact_billing_dao.py]: yeah I agree with you but I'm concerend that on the org dashboard we refer to `sms_billable_units`. That comes from `fetch_usage_year_for_organisation`, which calls `fetch_sms_billing_for_organisation`, which calls this function.

i think best option might be to make it so that it stays consistent as chargeable until it gets to the dict conversion (since that feels like output formatting and manipulation rather than data gathering). thoughts?

File: tests/app/dao/test_fact_billing_dao.py [hunk 2]
```diff
@@ -1127,6 +1121,37 @@ def test_query_organisation_sms_usage_for_year_handles_multiple_rates(notify_db_
     assert result[1]['cost'] == 6
 
 
+@freeze_time('2022-05-01 13:30')
+def test_query_organisation_sms_usage_for_year_filters_on_organisation(notify_db_session):
```
[the remark this answers is not in the document]
Comment (leohemsted) [reply tests/app/dao/test_fact_billing_dao.py]: Good shout! I've inlined it in 0955741. However, now I don't think we need a new test at all as `test_fetch_usage_year_for_organisation` asserts implicitly that we're only getting results for org_1 by virtue of the services returned in the results.

File: app/dao/fact_billing_dao.py [hunk 3]
```diff
@@ -70,11 +70,13 @@ def fetch_sms_billing_for_all_services(start_date, end_date):
     # ASSUMPTION: AnnualBilling has been populated for year.
     ft_billing_subquery = query_sms_usage_for_year_per_service(financial_year).subquery()
 
-    sms_billable_units = func.sum(func.coalesce(ft_billing_subquery.c.chargeable_units, 0))
+    sms_billable_units = func.sum(ft_billing_subquery.c.chargeable_units)
```
[the remark this answers is not in the document]
Comment (leohemsted) [reply app/dao/fact_billing_dao.py]: the coalesces are now back in. they ended up getting removed in the commit cos i'd removed them when debugging the raw sql to make it easier for me to read.

i think we need the coalesce on `sms_cost` because we say [`if float(s.sms_cost) > 0`](https://github.com/alphagov/notifications-api/blob/master/app/platform_stats/rest.py#L88) but we probably don't need it on any of the other points

File: app/dao/fact_billing_dao.py [hunk 4]
```diff
@@ -64,53 +64,52 @@ def fetch_sms_free_allowance_remainder_until_date(end_date):
...
+    sms_cost = func.sum(ft_billing_subquery.c.cost)
     query = db.session.query(
         Organisation.name.label('organisation_name'),
         Organisation.id.label('organisation_id'),
+
         Service.name.label("service_name"),
         Service.id.label("service_id"),
-        allowance_left_at_start_date_query.c.free_sms_fragment_limit,
-        FactBilling.rate.label('sms_rate'),
+        AnnualBilling.free_sms_fragment_limit,
```
[the remark this answers is not in the document]
Comment (benthorner) [reply app/dao/fact_billing_dao.py]: I think it's fine personally. We have other values coming from the subquery so it's not unusual - you could even go further and get `service_id` from it. If someone is trying to understand the function in depth, they'll need to understand the subquery anyway so I don't think it'll have a major impact. Conversely, I find joins much harder to reason about because of the filtering affect they can have, so avoiding them is helpful for me.

File: app/dao/fact_billing_dao.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/dao/fact_billing_dao.py]: Hmm...I'm not sure. Is there a test that can expose which coalesces we actually need?

Looking at the query, the coalesces might be necessary if a service doesn't have any rows in `ft_billing` i.e. "Missing: no data for a service (no row in results)". [We do have such a test for the organisation report](https://github.com/alphagov/notifications-api/blob/e604385e0cf4c2ab8c6451b7120ceb196cce21b5/tests/app/dao/test_fact_billing_dao.py#L942), but that code works in a very different way so I think it's a stretch to say it's covered already.

File: tests/app/dao/test_fact_billing_dao.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply tests/app/dao/test_fact_billing_dao.py]: Nice, and yugh [I see what you mean](https://github.com/alphagov/notifications-api/blob/e604385e0cf4c2ab8c6451b7120ceb196cce21b5/tests/app/db.py#L997).

File: app/dao/fact_billing_dao.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (leohemsted) [reply app/dao/fact_billing_dao.py]: if a service doesn't have any ft_billing then for the platform admin report they won't be returned at all - it only returns those services with non-zero data.

File: tests/app/dao/test_fact_billing_dao.py [hunk 5]
```diff
@@ -1095,15 +1092,12 @@ def test_query_organisation_sms_usage_for_year_handles_multiple_services(notify_
 
 
 @freeze_time('2022-05-01 13:30')
-def test_query_organisation_sms_usage_for_year_handles_multiple_rates(notify_db_session):
+def test_query_sms_usage_for_year_per_service_handles_multiple_rates(notify_db_session):
```
[the remark this answers is not in the document]
Comment (leohemsted) [reply tests/app/dao/test_fact_billing_dao.py:1107]: I've done the work but I don't think I like it - the query_sms_usage function is also used by `fetch_monthly_billing_for_year` and `fetch_billing_totals_for_year`, and given how complex it is i think these tests serve quite well as documentation, as a step by step breakdown of what's actually happening

either way i made the change so i pushed it up in d54bf4b2 for you to see - what do you think?

File: app/dao/fact_billing_dao.py [hunk 6]
```diff
@@ -64,53 +64,52 @@ def fetch_sms_free_allowance_remainder_until_date(end_date):
...
+    chargeable_sms = func.sum(ft_billing_subquery.c.charged_units)
+    sms_cost = func.sum(ft_billing_subquery.c.cost)
     query = db.session.query(
         Organisation.name.label('organisation_name'),
         Organisation.id.label('organisation_id'),
+
         Service.name.label("service_name"),
         Service.id.label("service_id"),
-        allowance_left_at_start_date_query.c.free_sms_fragment_limit,
-        FactBilling.rate.label('sms_rate'),
```
[the remark this answers is not in the document]
Comment (leohemsted) [reply app/dao/fact_billing_dao.py:90]: that's a fair point - i did leave a comment in the commit message 1118449150a07c1772a31d288b71565c71931208 which i hoped would cover it

File: app/dao/fact_billing_dao.py [hunk 7]
```diff
@@ -798,12 +802,15 @@ def query_sms_usage_for_year_per_service(year):
     # for, after taking any remaining free allowance into account.
     charged_units = func.greatest(this_rows_chargeable_units - remaining_free_allowance_before_this_row, 0)
 
+    free_allowance_used_to_date = chargeable_units_used_before_this_row + this_rows_chargeable_units
```
[the remark this answers is not in the document]
Comment (benthorner) [reply app/dao/fact_billing_dao.py]: Agreed this is sensible.

**Comment by benthorner** [comment]
Update: we now have follow-up PRs for the bulk of the work here.

- https://github.com/alphagov/notifications-api/pull/3559
- https://github.com/alphagov/notifications-api/pull/3561

**Comment by leohemsted** [comment]
Closing in favour of the PRs ben mentioned above
